Week3/HW03.hs

```diff
@@ -0,0 +1,169 @@
+
+data Expression =
+    Var String                   -- Variable
+  | Val Int                      -- Integer literal
+  | Op Expression Bop Expression -- Operation
+  deriving (Show, Eq)
+
+-- Binary (2-input) operators
+data Bop = 
+    Plus     
+  | Minus    
+  | Times    
+  | Divide   
+  | Gt
+  | Ge       
+  | Lt  
+  | Le
+  | Eql
+  deriving (Show, Eq)
+
+data Statement =
+    Assign   String     Expression
+  | Incr     String
+  | If       Expression Statement  Statement
+  | While    Expression Statement       
+  | For      Statement  Expression Statement Statement
+  | Sequence Statement  Statement        
+  | Skip
+  deriving (Show, Eq)
+
+type State = String -> Int 
+
+-- Exercise 1 -----------------------------------------
+
+extend :: State -> String -> Int -> State
+extend initState name val a             = bool (initState a) val (a == name)
+
+empty :: State
+empty _                                 = 0
+
+-- Exercise 2 -----------------------------------------
+
+toInt :: Bool -> Int
+toInt b = if b then 1 else 0 
+
+evalE :: State -> Expression -> Int
+evalE _ (Val a) = a 
+evalE initState (Op exp1 bop exp2) = 
+  let a = evalE initState exp1 
+      b = evalE initState exp2
+  in case bop 
+    of  Plus            -> a + b
+        Minus           -> a - b
+        Times           -> a * b
+        Divide          -> a `div` b
+        Gt              -> toInt $ a > b
+        Ge              -> toInt $ a >= b
+        Lt              -> toInt $ a < b
+        Le              -> toInt $ a <= b 
+        Eql             -> toInt $ a == b
+evalE _ _ = 0
+
+-- Exercise 3 -----------------------------------------
+
+data DietStatement = DAssign String Expression
+                   | DIf Expression DietStatement DietStatement
+                   | DWhile Expression DietStatement
+                   | DSequence DietStatement DietStatement
+                   | DSkip
+                     deriving (Show, Eq)
+
+desugar :: Statement -> DietStatement
+desugar (Assign s e)                    = DAssign s e 
+desugar (Incr s)                        = DAssign s (Op (Var s) Plus (Val 1))
+desugar (If exp1 st1 st2)               = DIf exp1 (desugar st1) (desugar st2)
+desugar (Sequence st1 st2)              = DSequence (desugar st1) (desugar st2)
+desugar (While exp1 st1)                = DWhile exp1 (desugar st1)
+desugar (For ini cond incr body)        = DSequence (desugar ini) (DWhile cond (DSequence (desugar body) (desugar incr)))
+desugar _                               = undefined
+
+-- Exercise 4 -----------------------------------------
+
+evalSimple :: State -> DietStatement -> State
+evalSimple initState (DAssign name exp1)       = extend initState name (evalE initState exp1)
+evalSimple initState (DIf check st1 st2)       = evalSimple initState (if evalE initState check == 1 then st1 else st2)
+evalSimple initState (DWhile exp1 st1)         = if evalE initState exp1 == 1 then evalSimple initState st1 else initState
```

Comment:
This implementation of while is not correct. While is supposed to run repeatedly as long as the condition is satisfied. You're running it just once in that case. Also don't forget to pass the modified state into the next iteration of the loop.

Reply:
thats why i wasnt getting the correct result for factorial and i thought i was messing something in writing the function call(test i called it).